### ISSUE TITLE
GEMMA models are not downloading for SANA workflow.

### ISSUE THREAD

**Issue by parthwagh9999** [city96/ComfyUI_ExtraModels#94]
![image](https://github.com/user-attachments/assets/15f01eca-6d71-4af2-9bf5-a968aadbccf7)

I tried downloading without comfy Ui, means, I downloaded through hugging face manually. 
Even though it is not recognizing existing download. It is downloading again.

Don't know how to add access token. When added access token through import method it is breaking the nodes.

**Comment by Raphael-Screenworks** [comment]
where do you place the gemma model inside comfy ui?
models/... ?

**Comment by parthwagh9999** [comment]
> where do you place the gemma model inside comfy ui? models/... ?

H:\ComfyUI_windows_portable\ComfyUI\models\text_encoders\models--google--gemma-2-2b-it

I made changes to gemma.py file by adding access token in snapshot download function. Then it started downloading this model,
but issue is persistent without it and not checking even when models are downloaded there.

Thank you

**Comment by parthwagh9999** [comment]
> > where do you place the gemma model inside comfy ui? models/... ?
> 
> H:\ComfyUI_windows_portable\ComfyUI\models\text_encoders\models--google--gemma-2-2b-it
> 
> I made changes to gemma.py file by adding access token in snapshot download function. Then it started downloading this model, but issue is persistent without it and not checking even when models are downloaded there.
> 
> Thank you

it again giving me error after some time, even using access token.

**Comment by city96** [comment]
Sana team uploaded a copy of the gemma2 weights that don't need an access token to download, it should be available in the dropdown now. https://github.com/city96/ComfyUI_ExtraModels/commit/4a770ac22b85e70de5bafd5f6b68118dddcef2fd

**Comment by parthwagh9999** [comment]
> Sana team uploaded a copy of the gemma2 weights that don't need an access token to download, it should be available in the dropdown now. [4a770ac](https://github.com/city96/ComfyUI_ExtraModels/commit/4a770ac22b85e70de5bafd5f6b68118dddcef2fd)

![image](https://github.com/user-attachments/assets/96062021-3b05-47fa-821c-4a9ca2ce2d39)
gave this error, because there is not path mentioned in gemma/node.py file.

add like this 

![image](https://github.com/user-attachments/assets/c1ba055e-4fab-4caf-bcc8-744eb6b906fa)

Now this started downloading model for gemma.

**Comment by parthwagh9999** [comment]
> That's what I get for editing stuff on 4 hours of sleep lol [6e5ad55](https://github.com/city96/ComfyUI_ExtraModels/commit/6e5ad55f400e3fcbf5d85fe54b0aa926a2d95a11)
> 
> At least I got the refactor halfway working which just lets you select actual local files instead of the auto download logic.
> 
> ![image](https://private-user-images.githubusercontent.com/125218114/395655556-9aa42b9d-a04d-4d35-95ba-e09370edb64d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQxNTQ0OTEsIm5iZiI6MTczNDE1NDE5MSwicGF0aCI6Ii8xMjUyMTgxMTQvMzk1NjU1NTU2LTlhYTQyYjlkLWEwNGQtNGQzNS05NWJhLWUwOTM3MGVkYjY0ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNFQwNTI5NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YzhhMGE4OTk1OGM5OTA3OTQ2NjZlY2NiNDQ4NzEwZDIxMTRjZDRmMTZhYjI1NjIzODhiMjQ4YzVmYjhiNGMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OIYVn1savdreOT0r6W8c7_KsSXtMFOgAszySRsFAeXQ)

Thank you so much for your valuable time. I understand, it happens to everyone.

**Comment by parthwagh9999** [comment]
Even these changes are not working. I tried running several times. Model is downloading each time. Even after downloading it, it is not running. I am getting black images. Your workflow is completely different from workflow provided in this page "https://github.com/city96/ComfyUI_ExtraModels?tab=readme-ov-file".  
Also, after updating extra model node, nodes are totally different from yours. Nodes mentioned in this image 
![image](https://github.com/user-attachments/assets/81ec40da-e610-445f-87be-784805bca267)
are not found on updated extra model node. 
![image](https://github.com/user-attachments/assets/ac3a3c0d-3169-47db-96d4-eaf9c9461847)
![image](https://github.com/user-attachments/assets/a191fbfb-61e0-4d9e-928b-7136e0f4ebdd)
![image](https://github.com/user-attachments/assets/37eabda7-d3e1-49ae-83ef-b1d7150c29b4)
![image](https://github.com/user-attachments/assets/7787f72a-84db-466e-8469-e8a78c04e93b)
![image](https://github.com/user-attachments/assets/63695def-f5c0-427d-9bb6-a5affccc5ece)

Even manually downloading models is not working. making changes to code like this to not download model, and use downloaded model giving this result 
![image](https://github.com/user-attachments/assets/9f5f8c73-cedd-44da-982a-be3026ec6594)

"import os
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

# Define folder_paths for model directories
class folder_paths:
    models_dir = "./models"  # Adjust to your model directory path
    folder_names_and_paths = {
        "text_encoders": os.path.join(models_dir, "text_encoders")
    }

from ..utils.dtype import string_to_dtype

# Root directory for manually downloaded text encoder models
tenc_root = (
    folder_paths.folder_names_and_paths.get(
        "text_encoders",
        folder_paths.folder_names_and_paths.get("clip", [[], set()])
    )
)

dtypes = [
    "default",
    "auto (comfy)",
    "BF16",
    "FP32",
    "FP16",
]

try:
    torch.float8_e5m2
except AttributeError:
    print("Torch version too old; FP8 not supported")
else:
    dtypes += ["FP8 E4M3", "FP8 E5M2"]

class GemmaLoader:
    @classmethod
    def INPUT_TYPES(s):
        devices = ["auto", "cpu", "cuda"]
        # Support multiple GPUs
        for k in range(1, torch.cuda.device_count()):
            devices.append(f"cuda:{k}")
        return {
            "required": {
                "model_name": ([
                    "Efficient-Large-Model/gemma-2-2b-it",
                    "google/gemma-2-2b-it",
                    "unsloth/gemma-2-2b-it-bnb-4bit"
                ],),
                "device": (devices, {"default": "cpu"}),
                "dtype": (dtypes,),
            }
        }

    RETURN_TYPES = ("GEMMA",)
    FUNCTION = "load_model"
    CATEGORY = "ExtraModels/Gemma"
    TITLE = "Gemma Loader"

    def load_model(self, model_name, device, dtype):
        dtype = string_to_dtype(dtype, "text_encoder")
        if device == "cpu":
            assert dtype in [None, torch.float32], f"Can't use dtype '{dtype}' with CPU! Set dtype to 'default'."

        # Define the local directory for the manually downloaded model
        if model_name == 'google/gemma-2-2b-it':
            text_encoder_dir = text_encoder_dir = r"H:\ComfyUI_windows_portable\ComfyUI\models\text_encoders\models--google--gemma-2-2b-it"
        elif model_name == 'unsloth/gemma-2-2b-it-bnb-4bit':
            text_encoder_dir = os.path.join(folder_paths.models_dir, 'text_encoders', 'models--unsloth--gemma-2-2b-it-bnb-4bit')
        elif model_name == 'Efficient-Large-Model/gemma-2-2b-it':
            text_encoder_dir = os.path.join(folder_paths.models_dir, 'text_encoders', 'models--Efficient-Large-Model--gemma-2-2b-it')
        else:
            raise ValueError('Model not implemented!')

        # Check if the manually downloaded model exists
        if not os.path.exists(text_encoder_dir) or not os.listdir(text_encoder_dir):
            raise FileNotFoundError(f"Manually downloaded model files not found in: {text_encoder_dir}")

        # Load tokenizer and model from the local directory
        tokenizer = AutoTokenizer.from_pretrained(text_encoder_dir)
        text_encoder_model = AutoModelForCausalLM.from_pretrained(text_encoder_dir, torch_dtype=dtype)
        tokenizer.padding_side = "right"
        text_encoder = text_encoder_model.get_decoder()

        if device != "cpu":
            text_encoder = text_encoder.to(device)

        return ({
            "tokenizer": tokenizer,
            "text_encoder": text_encoder,
            "text_encoder_model": text_encoder_model
        },)

class GemmaTextEncode:
    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "text": ("STRING", {"multiline": True}),
                "GEMMA": ("GEMMA",),
            }
        }

    RETURN_TYPES = ("CONDITIONING",)
    FUNCTION = "encode"
    CATEGORY = "ExtraModels/Gemma"
    TITLE = "Gemma Text Encode"

    def encode(self, text, GEMMA=None):
        print(text)
        tokenizer = GEMMA["tokenizer"]
        text_encoder = GEMMA["text_encoder"]

        with torch.no_grad():
            tokens = tokenizer(
                text,
                max_length=300,
                padding="max_length",
                truncation=True,
                return_tensors="pt"
            ).to(text_encoder.device)

            cond = text_encoder(tokens.input_ids, tokens.attention_mask)[0]
            emb_masks = tokens.attention_mask

        cond = cond * emb_masks.unsqueeze(-1)

        return ([[cond, {}]], )

NODE_CLASS_MAPPINGS = {
    "GemmaLoader": GemmaLoader,
    "GemmaTextEncode": GemmaTextEncode,
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "GemmaLoader": "Gemma Loader",
    "GemmaTextEncode": "Gemma Text Encode",
}
![image](https://github.com/user-attachments/assets/7c1249ef-0979-4f22-8481-b16d7170d899)
"

**Comment by jonnytracker** [comment]
401 Client Error. (Request ID: Root=1-675d8b33-604123b6654831703760a42c;a38b7bcf-24dc-4b79-bc8b-eb68f613c5b2)

Cannot access gated repo for url https://huggingface.co/google/gemma-2-2b-it/resolve/299a8560bedf22ed1c72a8a11e7dce4a7f9f51f8/.gitattributes.
Access to model google/gemma-2-2b-it is restricted. You must have access to it and be authenticated to access it. Please log in.

**Comment by jasoncow007** [comment]
> > That's what I get for editing stuff on 4 hours of sleep lol [6e5ad55](https://github.com/city96/ComfyUI_ExtraModels/commit/6e5ad55f400e3fcbf5d85fe54b0aa926a2d95a11)
> > At least I got the refactor halfway working which just lets you select actual local files instead of the auto download logic.
> > ![image](https://private-user-images.githubusercontent.com/125218114/395655556-9aa42b9d-a04d-4d35-95ba-e09370edb64d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQxNTQ0OTEsIm5iZiI6MTczNDE1NDE5MSwicGF0aCI6Ii8xMjUyMTgxMTQvMzk1NjU1NTU2LTlhYTQyYjlkLWEwNGQtNGQzNS05NWJhLWUwOTM3MGVkYjY0ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNFQwNTI5NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YzhhMGE4OTk1OGM5OTA3OTQ2NjZlY2NiNDQ4NzEwZDIxMTRjZDRmMTZhYjI1NjIzODhiMjQ4YzVmYjhiNGMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OIYVn1savdreOT0r6W8c7_KsSXtMFOgAszySRsFAeXQ)
> 
> Thank you so much for your valuable time. I understand, it happens to everyone.

cannot recognize local path, is there a cliploader node just like in your picture

**Comment by city96** [comment]
That image is from the refactor branch here, but it's not 1:1 to the reference version yet https://github.com/city96/ComfyUI_ExtraModels/pull/92

**Comment by jasoncow007** [comment]
> That image is from the refactor branch here, but it's not 1:1 to the reference version yet #92

swich to rewrite tag, the samperling is terribly slow, any suggestion?

**Comment by city96** [comment]
@jasoncow007 you need the empty sana latent node for the latent, otherwise you're generating an image 4x the size of what you actually want, which explains why it would be extremely slow.

**Comment by jasoncow007** [comment]
> @city96  you need the empty sana latent node for the latent, otherwise you're generating an image 4x the size of what you actually want, which explains why it would be extremely slow.
![image](https://github.com/user-attachments/assets/8dae5c50-d309-4043-967a-2891624ca3e6)

change to sana latent, it works,but only the sampling, got the black image at last, any suggestion

**Comment by jasoncow007** [comment]
> > Sana team uploaded a copy of the gemma2 weights that don't need an access token to download, it should be available in the dropdown now. [4a770ac](https://github.com/city96/ComfyUI_ExtraModels/commit/4a770ac22b85e70de5bafd5f6b68118dddcef2fd)
> 
> ![image](https://private-user-images.githubusercontent.com/59022259/395642366-96062021-3b05-47fa-821c-4a9ca2ce2d39.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwMjkzMjUsIm5iZiI6MTczNTAyOTAyNSwicGF0aCI6Ii81OTAyMjI1OS8zOTU2NDIzNjYtOTYwNjIwMjEtM2IwNS00N2ZhLTgyMWMtNGE5Y2EyY2UyZDM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI0VDA4MzAyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwMjcxOGE4NmM4NjU5OGM2NjJiZDY4MWFlMjAwMDE2MDZhNWQ2ZWI1Nzg2MzlmOTdhODc2NDgzZTNkN2U4ZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dTJNKPxu65HysAlEP1muJi7aVZ2VHjtKQH2004bdgR8) gave this error, because there is no path mentioned in gemma/node.py file.
> 
> add like this
> 
> ![image](https://private-user-images.githubusercontent.com/59022259/395643051-c1ba055e-4fab-4caf-bcc8-744eb6b906fa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwMjkzMjUsIm5iZiI6MTczNTAyOTAyNSwicGF0aCI6Ii81OTAyMjI1OS8zOTU2NDMwNTEtYzFiYTA1NWUtNGZhYi00Y2FmLWJjYzgtNzQ0ZWI2YjkwNmZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI0VDA4MzAyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5NGQxMWYxNTBmNmUyMmRiMWYwZWNmMmNiY2Q2MThlYjQwYzUwODQ2NWQxNzA5MzEwYjFmM2FlYzdiMmMyZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.R1jj6-_jMgYP1ett1TS9-xLDSOYNivuy85eZajA7e4k)
> 
> Now this started downloading model for gemma.

hi,buddy! have you done with this issue? it still cannot recognize local path here
